lib/charms/grafana_agent/v0/cos_machine.py

```diff
@@ -0,0 +1,257 @@
+        return [
+            {"job_name": f"{job_name_prefix}_{key}", **endpoint}
+            for key, endpoint in enumerate(self._metrics_endpoints)
+        ]
+
+    @property
+    def _metrics_alert_rules(self) -> Dict:
+        """Use (for now) the prometheus_scrape AlertRules to initialize this."""
+        alert_rules = MetricsAlerts()
+        alert_rules.add_path(self._metrics_rules, recursive=self._recursive)
+        return alert_rules.as_dict()
+
+    @property
+    def _log_alert_rules(self) -> Dict:
+        """Use (for now) the loki_push_api AlertRules to initialize this."""
+        alert_rules = LogAlerts()
+        alert_rules.add_path(self._logs_rules, recursive=self._recursive)
+        return alert_rules.as_dict()
+
+    @property
+    def _dashboards(self) -> List[str]:
+        dashboards = []
+        for d in self._dashboard_dirs:
+            for path in Path(d).glob("*"):
+                dashboards.append(self._encode_dashboard_content(path.read_bytes()))
+
+        return dashboards
+
+    def _encode_dashboard_content(self, content: Union[str, bytes]) -> str:
+        if isinstance(content, str):
+            content = bytes(content, "utf-8")
+
+        return base64.b64encode(lzma.compress(content)).decode("utf-8")
+
+
+class COSMachineDataChanged(EventBase):
+    """Event emitted when a `COSMachineProvider` joins or updates data."""
+
+
+class COSMachineConsumerEvents(ObjectEvents):
+    """Event descriptor for events raised by `COSMachineConsumer`."""
+
+    data_changed = EventSource(COSMachineDataChanged)
+
+
+class COSMachineConsumer(Object):
+    """Provider class for the cos_machine interface."""
+
+    on = COSMachineConsumerEvents()
+
+    def __init__(
+        self,
+        charm,
+        relation_name: str = DEFAULT_RELATION_NAME,
+        refresh_events: Optional[List] = None,
+    ):
+        """Create a COSMachineConsumer instance.
+
+        Args:
+            charm: The `CharmBase` instance that is instantiating this object.
+            relation_name: The name of the relation to communicate over.
+            refresh_events: List of events on which to resfresh relation data.
+        """
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+        self._refresh_events = refresh_events or [self._charm.on.config_changed]
+
+        events = self._charm.on[relation_name]
+        self.framework.observe(events.relation_joined, self._on_relation_data_changed)
+        self.framework.observe(events.relation_changed, self._on_relation_data_changed)
+        for event in self._refresh_events:
+            self.framework.observe(event, self.trigger_refresh)
+
+    def _on_relation_data_changed(self, _):
+        self.on.data_changed.emit()
+
+    def trigger_refresh(self, _):
+        """Trigger a refresh of relation data."""
+        # FIXME: Figure out what we should do here
+        self.on.data_changed.emit()
```

Comment:
I don't see the point of this. The charm would just call this to trigger data_changed? Why not just call the handler of data_changed instead?

Reply:
Yeah this doesn't look like charm-facing API.

Reply:
I see the value of having refresh events as mentioned by @rbarry82 elsewhere, but I think we can manage by merging `_on_relation_data_changed` and `trigger_refresh` in one private method, that the lib uses when it's time to signal to the owner charm that it's time for a refresh.

Maybe it's good to rename `data_changed` to `refresh`, because `changed` might give the wrong impression that there's a delta, while I don't think that is in fact necessarily the case.